torch/csrc/distributed/rpc/python_functions.cpp

```diff
@@ -134,42 +134,40 @@ std::shared_ptr<FutureMessage> sendPythonRemoteCall(
 
 using namespace torch::distributed::autograd;
 
-c10::intrusive_ptr<JitFuture> wrapFutureMessageInJitFuture(
-    const std::shared_ptr<FutureMessage>& futureResponseMessage,
+c10::intrusive_ptr<JitFuture> toPyJitFuture(
+    const std::shared_ptr<JitFuture>& messageJitFuture,
     bool hasValue) {
   if (hasValue) {
-    c10::intrusive_ptr<JitFuture> jitFuture =
+    c10::intrusive_ptr<JitFuture> pyJitFuture =
         c10::make_intrusive<JitFuture>(PyObjectType::get());
-    std::weak_ptr<FutureMessage> wp = futureResponseMessage;
-    futureResponseMessage->addCallback(
-        at::wrapPropagateTLSState<void>([jitFuture, wp]() {
-          auto futureResponseMessage = wp.lock();
-          if (futureResponseMessage->hasError()) {
-            jitFuture->setError(
-                std::make_exception_ptr(*futureResponseMessage->error()));
+    std::weak_ptr<JitFuture> wp = messageJitFuture;
+    messageJitFuture->addCallback(
+        at::wrapPropagateTLSState<void>([pyJitFuture, wp]() {
+          auto messageJitFuture = wp.lock();
+          if (messageJitFuture->hasError()) {
+            pyJitFuture->setError(messageJitFuture->exception_ptr());
           } else {
-            jitFuture->markCompleted(
-                toIValue(futureResponseMessage->constValue()));
+            pyJitFuture->markCompleted(
+                toIValue(*messageJitFuture->value().toCustomClass<Message>()));
```

Comment:
I'm not familiar with JIT intricacies, but to me this looks like we're casting the IValue to a Message instance, just to then convert it back to an IValue. What am I missing?

Reply:
Another thing I'm not getting: the future we return has type `PyObject` but it still contains the same `Message` instance. Does this mean we're exposing Message to Python?

Reply:
The input IValue contains a message, and the output IValue contains a PyObject. The `toIValue` function deserializes the message based on the type and convert that into `py::object` and then wrap it into an IValue. Let me add more comments to clarify that.

Reply:
I see, I guess I just assumed what `toIValue` did without bothering to go check.